proposals/mesh-detection.html

```diff
@@ -0,0 +1,450 @@
+        
+        // set up three.js boilerplate
+        scene = new THREE.Scene();
+
+        camera = new THREE.PerspectiveCamera(70, window.innerWidth / window.innerHeight, 0.01, 20);
+
+        const light = new THREE.HemisphereLight(0xffffff, 0xbbbbff, 1);
+        light.position.set(0.5, 1, 0.25);
+        scene.add(light);
+
+        renderer = new THREE.WebGLRenderer({ antialias: true, alpha: true });
+        renderer.setPixelRatio(window.devicePixelRatio);
+        renderer.setSize(window.innerWidth, window.innerHeight);
+        renderer.xr.enabled = true;
+        renderer.xr.setFoveation(0);
+        renderer.autoClear = false;
+        container.appendChild(renderer.domElement);
+
+        xrButton = new WebXRButton({
+          onRequestSession: onRequestSession,
+          onEndSession: onEndSession,
+          textEnterXRTitle: "START AR",
+          textXRNotFoundTitle: "AR NOT FOUND",
+          textExitXRTitle: "EXIT AR",
+        });
+
+        document.querySelector('header').appendChild(xrButton.domElement);
+
+        if (navigator.xr) {
+          navigator.xr.isSessionSupported('immersive-ar')
+                      .then((supported) => {
+            xrButton.enabled = supported;
+          });
+        }
+
+        // add controllers to the scene
+				controller1 = renderer.xr.getController(0);
+				controller1.addEventListener('selectstart', onSelectStart);
+				controller1.addEventListener('selectend', onSelectEnd);
+				controller1.addEventListener('connected', function (event) {
+					this.add(buildController(event.data));
+				});
+
+				controller1.addEventListener('disconnected', function () {
+					this.remove(this.children[ 0 ]);
+				});
+
+				scene.add(controller1);
+
+				controller2 = renderer.xr.getController(1);
+				controller2.addEventListener('selectstart', onSelectStart);
+				controller2.addEventListener('selectend', onSelectEnd);
+				controller2.addEventListener('connected', function (event) {
+					this.add(buildController(event.data));
+				});
+
+        scene.add(controller2);
+
+				const controllerModelFactory = new XRControllerModelFactory();
+
+				controllerGrip1 = renderer.xr.getControllerGrip(0);
+				controllerGrip1.add(controllerModelFactory.createControllerModel(controllerGrip1));
+				scene.add(controllerGrip1);
+
+				controllerGrip2 = renderer.xr.getControllerGrip(1);
+				controllerGrip2.add(controllerModelFactory.createControllerModel(controllerGrip2));
+				scene.add(controllerGrip2);
+
+				controller2.addEventListener('disconnected', function () {
+					this.remove(this.children[ 0 ]);
+				});
+				scene.add(controller2);
+
+        reticle = new THREE.Mesh(
+          new THREE.SphereGeometry(0.15,32,16),
+          new THREE.MeshBasicMaterial({color: 0xff0000})
+        );
+
+        reticle.matrixAutoUpdate = true;
+        reticle.visible = false;
+        scene.add(reticle);
+
+        const createMeshMaterial = (params) => 
```

Comment:
Is there a reason this method needs to be duplicated?

Reply:
no, I moved it to a helper function